sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,123 +32,206 @@
 
 import org.joda.time.Instant;
 
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+
 /**
  * Determine the timing and other properties of a new pane for a given computation, key and window.
  * Incorporates any previous pane, whether the pane has been produced because an
  * on-time {@link AfterWatermark} trigger firing, and the relation between the element's timestamp
  * and the current output watermark.
  */
 public class PaneInfoTracker {
-  private TimerInternals timerInternals;
 
-  public PaneInfoTracker(TimerInternals timerInternals) {
-    this.timerInternals = timerInternals;
+  /**
+   * Interface for calculating and tracking pane info state.
+   */
+  public interface PaneInfoTrackerIface {
```

Comment:
It turned out to work much better if I went back to the single class and use the window maxTimestamp to distinguish per element.

What tipped me over the edge was the check for GlobalWindows.INSTANCE.isCompatible(...) fails for the IdentityWindowFn wrapping the/a GlobalWindows. Rather than relying on static types, I convinced myself any window who's end is at or beyond GlobalWindow.INSTANCE.maxTimestamp is 'open forever' and thus subject to all the anti-state leaking fixes here.